### PR TITLE
basic tests for apps.table

### PR DESCRIPTION
**Pull request by 0xmilkmix**
This PR is related to https://github.com/osquery/osquery/issues/5020

I enabled the test for the apps table and specified some of the fields to be more detailed than `NormalType`

Fixes https://github.com/osquery/osquery/issues/5020

### CODE REVIEW

**Comment by directionless** [comment]
Thanks for doing this! It looks like it's failing on the format linting. From a build directory you should be able to `make format_check`

**Comment by 0xmilkmix** [comment]
> Thanks for doing this! It looks like it's failing on the format linting. From a build directory you should be able to `make format_check`

sorry for the commit error, this should be fixed

**Comment by 0xmilkmix** [comment]
could it be that some tests work on my laptop but not on AzureEnv? 2 columns were working locally but not in the CI.
Also, `StartupItemsTest` `username` test is failing locally but does not seems to in Azure

**Comment by theopolis** [comment]
@0xmilkmix, we should investigate the differences. Do you have more details about what is failing locally for you?

**Comment by theopolis** [comment]
I see in the previous CI output:
```
0: Standard validator of the column "element" with value "" failed
90: /Users/runner/runners/2.159.2/work/1/s/tests/integration/tables/helper.cpp:160: Failure
90: Value of: validate_value_using_flags(value, flags)
90:   Actual: false
90: Expected: true
90: Standard validator of the column "applescript_enabled" with value "" failed
90: /Users/runner/runners/2.159.2/work/1/s/tests/integration/tables/helper.cpp:160: Failure
90: Value of: validate_value_using_flags(value, flags)
90:   Actual: false
90: Expected: true
```

**Comment by 0xmilkmix** [comment]
Currently, I've the following test failing locally (not from my MR) but from the CI on Azure, everything is ok

```
90: [----------] 1 test from StartupItemsTest
90: [ RUN      ] StartupItemsTest.test_sanity
90: <...>/osquery/tests/integration/tables/helper.cpp:160: Failure
90: Value of: validate_value_using_flags(value, flags)
90:   Actual: false
90: Expected: true
90: Standard validator of the column "username" with value "" failed
90: [  FAILED  ] StartupItemsTest.test_sanity (134 ms)
90: [----------] 1 test from StartupItemsTest (134 ms total)
```

File: tests/integration/tables/apps.cpp [hunk 1]
```diff
@@ -14,46 +14,38 @@
...
+      {"bundle_package_type", NormalType},
+      {"environment", NormalType},
+      {"element", NormalType},
+      {"compiler", NormalType},
+      {"development_region", NormalType},
+      {"display_name", NormalType},
+      {"info_string", NormalType},
+      {"minimum_system_version", NormalType},
+      {"category", NormalType},
+      {"applescript_enabled", NormalType},
```
Comment (theopolis) [inline tests/integration/tables/apps.cpp]: I think we want to address this. Right now it seems the values are either empty or `1`.

File: tests/integration/tables/apps.cpp [hunk 2]
```diff
@@ -14,46 +14,38 @@
...
+      {"name", NonEmptyString},
+      {"path", DirectoryOnDisk},
+      {"bundle_executable", NormalType},
+      {"bundle_identifier", NormalType},
+      {"bundle_name", NormalType},
+      {"bundle_short_version", NormalType},
+      {"bundle_version", NormalType},
+      {"bundle_package_type", NormalType},
+      {"environment", NormalType},
+      {"element", NormalType},
```
Comment (theopolis) [inline tests/integration/tables/apps.cpp]: This is either empty or `1`.

Comment (0xmilkmix) [reply tests/integration/tables/apps.cpp]: for `bundle_version` and `bundle_package_type` I've other values on my dev environment:
- version
  - `1.2.13.80` for GoogleSoftwareUpdateAgent.app
  - `3.7.4` for Python
- package_type
  - `BNDL`
  - `APPL`
for `environment` and `element` I indeed see only empty or `1`

File: tests/integration/tables/apps.cpp (hunk 1 above)
Comment (0xmilkmix) [reply tests/integration/tables/apps.cpp]: also have different values for those fields
- `minimum_system_version`
  - `10.7`
  - `10.14`
- `info_string`
  - `Eclipse 4.13 for Mac OS X, Copyright IBM Corp. and others 2002, 2019. All rights reserved.`
  - 3.7.4, © 2001-2019 Python Software Foundation
  - ...

**Comment by Smjert** [comment]
Please rebase on latest master, the macOS failure should be fixed.

**Comment by 0xmilkmix** [comment]
> Please rebase on latest master, the macOS failure should be fixed.

looks like rebasing now creates an error due to unsigned CLA

**Comment by Smjert** [comment]
> > Please rebase on latest master, the macOS failure should be fixed.
> 
> looks like rebasing now creates an error due to unsigned CLA

Please use `git rebase` to sync with master and then push force, don't use merge.

File: tests/integration/tables/apps.cpp [hunk 3]
```diff
@@ -10,50 +10,55 @@
...
-  //      {"minimum_system_version", NormalType}
-  //      {"category", NormalType}
-  //      {"applescript_enabled", NormalType}
-  //      {"copyright", NormalType}
-  //      {"last_opened_time", NormalType}
-  //}
-  // 4. Perform validation
-  // validate_rows(data, row_map);
+
+  auto verify_zero_one_or_empty = [](std::string const& value) {
```
Comment (directionless) [inline tests/integration/tables/apps.cpp]: Is this common enough it should be in helpers? It's would be `BoolOrEmpty` or something

**Comment by theopolis** [comment]
I recommend we fix the table (bool-or-empty feels weird).

**Comment by 0xmilkmix** [comment]
> I recommend we fix the table (bool-or-empty feels weird).

I can have a look. Sorry for the delay

**Comment by theopolis** [comment]
Heads up, placing this back into "needs response" since I think we need to look at the table and land the fix there (vs in the test).

**Comment by theopolis** [comment]
I'm going to close this PR. I think we should approach this differently and instead change the logic in the `apps` table.

**Pull request closed without merging**